modules/common/tls/tls.go

```diff
@@ -0,0 +1,235 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+)
+
+// This is just a quick fix, we can have this defined differently
+const MaxNumberOfCAs = 10
+
+// TLSService contains server-specific TLS settings
+type TLSService struct {
+	// Server-specific settings
+	SecretName             string `json:"secretName,omitempty"`
+	DisableNonTLSListeners bool   `json:"disableNonTLSListeners,omitempty"`
+}
+
+// TLSCa contains CA-specific settings, which could be used both by services (to define their own CA certificates)
+// and by clients (to verify the server's certificate).
+type TLSCa struct {
+	// CA-specific settings
+	CaSecretName string `json:"caSecretName,omitempty"`
+}
+
+// TLS is ageneric type, which encapsulates both the service and CA configurations
+type TLS struct {
+	Service TLSService `json:"service"`
+	Ca      TLSCa      `json:"ca"`
+}
+
+// NewTLS - initialize and return a TLS struct.
+func NewTLS(clientset *kubernetes.Clientset, namespace string, service TLSService, ca TLSCa) (*TLS, error) {
+	if service.SecretName == "" {
+		service.SecretName = "default-tls-secret"
+	} else {
+		err := EnsureOrCreateSecret(clientset, namespace, service.SecretName, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if ca.CaSecretName == "" {
+		ca.CaSecretName = "default-ca-secret"
+	} else {
+		err := EnsureOrCreateSecret(clientset, namespace, ca.CaSecretName, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &TLS{
+		Service: service,
+		Ca:      ca,
+	}, nil
+}
+
+// EnsureOrCreateSecret - checks if a secret with the given name exists in the specified namespace
+func EnsureOrCreateSecret(clientset *kubernetes.Clientset, namespace, secretName string, data map[string][]byte) error {
+	_, err := clientset.CoreV1().Secrets(namespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Secret not found, create it
+			secret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretName,
+					Namespace: namespace,
+				},
+				Data: data, // The data to store in the secret - the actual data provided by cert-manager formatted as "tls.crt": []byte("..."),  "tls.key": []byte("..."),
+			}
+			_, err := clientset.CoreV1().Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+			if err != nil {
+				return fmt.Errorf("Error creating secret %s: %w", secretName, err)
+			}
+		} else {
+			return fmt.Errorf("Error validating secret %s: %w", secretName, err)
+		}
+	}
+	return nil
+}
+
+// CreateVolumeMounts - add volume mount for TLS certificate and CA certificates
+func CreateVolumeMounts(tls *TLS) []corev1.VolumeMount {
+	var volumeMounts []corev1.VolumeMount
+
+	if tls.Service.SecretName != "" {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      "tls-certs",
+			MountPath: "/etc/tls-certs",
```

Comment:
I have some doubts on directly bind-mounting certs in /etc, because then the permission of the certs/ private keys and whatnot must be managed by this volumeMount.
We could bind-mount it under /var/{something}, and let Kolla config copy it in /etc/tls-certs and specify the  permissions in the Kolla config file? (for the record, this is how we use to do it with TripleO)

Reply:
Agree, changed to `/var/tls-certs` and /`var/ca-certs` instead.